connectivity.tf

```
@@ -23,9 +46,16 @@ resource "azurerm_virtual_network" "this" {
     for_each = each.value.subnets
 
     content {
-      name                            = subnet.value.name
-      address_prefixes                = subnet.value.address_prefixes
-      security_group                  = subnet.value.security_group
+      name             = subnet.value.name
+      address_prefixes = subnet.value.address_prefixes
+      // Use value from `var.connectivity.subnets.this.security_group_id
+      // fallback to resolved id of `var.connectivity.subnets.this.security_group_name`
+      // fallback to null.
+      // In essence 
+      // security_group_id > security_group_name > null
+      security_group = try(subnet.value.security_group_id, try(
```

Is it possible to use Coalesce?
Coalesce returns the first non-null value from left to right.
If all values are null, it returns the final fallback value, which in this case is null.
coalesce(
  subnet.value.security_group_id,
  azurerm_network_security_group.this["${each.key}_${subnet.key}"].id,
  null
)

---

It is possible but not in the way you describe. `coalesce` _fails_ when only non-null values are supplied, e.g.: `coalesce(null, null, null)`. `coalesce` always expect at least 1 value to be a non-null. However, that means we can simplify the expression in the referenced code with this. Here the `try` will swallow the failure of `coalesce` evaluating all null values.
```hcl
try(coalesce(
  subnet.value.security_group_id,
  azurerm_network_security_group.this["${each.key}_${subnet.key}"].id
), null)
```

Code example that proves `coalesce` fails on all null values:
```hcl
output "o" {
  value = coalesce(null, null, null)
}
```

Output:
```shell
╷
│ Error: Error in function call
│
│   on main.tf line 2, in output "o":
│    2:   value = coalesce(null, null, null)
│     ├────────────────
│     │ while calling coalesce(vals...)
│
│ Call to function "coalesce" failed: no non-null, non-empty-string arguments.
╵
```